storage/sealing.go

```diff
@@ -0,0 +1,253 @@
+
+	return out
+}
+
+func (m *Miner) sectorStateLoop(ctx context.Context) {
+	// TODO: restore state
+
+	go func() {
+		defer log.Warn("quitting deal provider loop")
+		defer close(m.stopped)
+
+		for {
+			select {
+			case sector := <-m.sectorIncoming:
+				m.onSectorIncoming(sector)
+			case update := <-m.sectorUpdated:
+				m.onSectorUpdated(ctx, update)
+			case <-m.stop:
+				return
+			}
+		}
+	}()
+}
+
+func (m *Miner) onSectorIncoming(sector *SectorInfo) {
+	has, err := m.sectors.Has(sector.SectorID)
+	if err != nil {
+		return
+	}
+	if has {
+		log.Warnf("SealPiece called more than once for sector %d", sector.SectorID)
+		return
+	}
+
+	if err := m.sectors.Begin(sector.SectorID, sector); err != nil {
+		// We may have re-sent the proposal
+		log.Errorf("deal tracking failed: %s", err)
+		m.failSector(sector.SectorID, err)
+		return
+	}
+
+	go func() {
+		select {
+		case m.sectorUpdated <- sectorUpdate{
+			newState: api.Packing,
+			id:       sector.SectorID,
+		}:
+		case <-m.stop:
+			log.Warn("failed to send incoming sector update, miner shutting down")
+		}
+	}()
+}
+
+func (m *Miner) onSectorUpdated(ctx context.Context, update sectorUpdate) {
+	log.Infof("Sector %d updated state to %s", update.id, api.SectorStateStr(update.newState))
+	var sector SectorInfo
+	err := m.sectors.Mutate(update.id, func(s *SectorInfo) error {
+		s.State = update.newState
+		if update.mut != nil {
+			update.mut(s)
+		}
+		sector = *s
+		return nil
+	})
+	if update.err != nil {
+		log.Errorf("sector %d failed: %s", update.id, update.err)
+		m.failSector(update.id, update.err)
+		return
+	}
+	if err != nil {
+		m.failSector(update.id, err)
+		return
+	}
+
+	switch update.newState {
+	case api.Packing:
+		m.handle(ctx, sector, m.finishPacking, api.Unsealed)
+	case api.Unsealed:
+		m.handle(ctx, sector, m.sealPreCommit, api.PreCommitting)
+	case api.PreCommitting:
+		m.handle(ctx, sector, m.preCommit, api.PreCommitted)
+	case api.PreCommitted:
+		m.handle(ctx, sector, m.preCommitted, api.SectorNoUpdate)
+	case api.Committing:
+		m.handle(ctx, sector, m.committing, api.Proving)
+	}
+}
+
+func (m *Miner) failSector(id uint64, err error) {
+	log.Errorf("sector %d error: %+v", id, err)
+	panic(err) // todo: better error handling strategy
+}
+
+func (m *Miner) SealPiece(ctx context.Context, ref string, size uint64, r io.Reader, dealID uint64) (uint64, error) {
+	log.Infof("Seal piece for deal %d", dealID)
+
+	sid, err := m.sb.AcquireSectorId() // TODO: Put more than one thing in a sector
+	if err != nil {
+		return 0, xerrors.Errorf("acquiring sector ID: %w", err)
+	}
+
+	ppi, err := m.sb.AddPiece(size, sid, r, []uint64{})
+	if err != nil {
+		return 0, xerrors.Errorf("adding piece to sector: %w", err)
+	}
+
+	return sid, m.newSector(ctx, sid, dealID, ref, ppi)
+}
+
+func (m *Miner) newSector(ctx context.Context, sid uint64, dealID uint64, ref string, ppi sectorbuilder.PublicPieceInfo) error {
+	si := &SectorInfo{
+		State:    api.UndefinedSectorState,
```

Comment:
undefined sector state? whats the purpose of doing this?

Reply:
`onIncoming` sets this to `Packing`after the sector arrives in state manager